portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,187 @@
+      if (!iterator.hasNext()) {
+        if (!currentRange.equals(lastExaminedTimeRange)) {
+          iterator.add(lastExaminedTimeRange);
+        }
+        // Once we, potentially, add the last element we should end the
+        // loop because it is the final range in the process.
+        break;
+      }
+    }
+  }
+
+  /**
+   * Helper method for merging two time ranges. This method is package-private so that it can be
+   * tested.
+   */
+  static TimeRange mergeTwoTimeRanges(TimeRange a, TimeRange b) {
+    if (!a.overlaps(b)) {
+      throw new IllegalArgumentException("Merging two time ranges that do not overlap is invalid");
+    }
+
+    Instant newTimeRangeStart;
+    Instant newTimeRangeEnd;
+
+    // The new time range after merging should have the earlier start time
+    // and the later end time among the two overlapping time ranges.
+    if (a.start().isBefore(b.start())) {
+      newTimeRangeStart = a.start();
+    } else {
+      newTimeRangeStart = b.start();
+    }
+
+    if (a.end().isBefore(b.end())) {
+      newTimeRangeEnd = b.end();
+    } else {
+      newTimeRangeEnd = a.end();
+    }
+
+    return TimeRange.fromStartEnd(newTimeRangeStart, newTimeRangeEnd);
+  }
+
+  /** Returns an iterator for the list of all time ranges. */
+  public Iterator<TimeRange> getAllTimeRangesIterator() {
+    return allTimeRanges.iterator();
+  }
+
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection. This method
+   * uses linear search to find the time ranges whose start time is before the target range's start
+   * and whose end time is after the target range's end. Then the method calls contains to see if
+   * the target range is contained within this current range.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
+
+    for (TimeRange currentRange : allTimeRanges) {
+      if (currentRange.contains(timeRangeToCheck)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Delete a time range from the list. Because the list of all time ranges are always kept to be
+   * pairwise disjoint, the potentially two new time ranges resulted from a deletion will not
+   * overlap with any other existing time ranges.
+   *
+   * <p>For example, if the list contains [3:00 - 4:00] and [5:00 - 6:00], deleting [3:15 - 3:30]
+   * will result in [3 - 3:15] and [3:30 - 4] as new time ranges.
+   *
+   * <p>Another example for deleting overlapping time ranges: if [3 - 4] and [5 - 6] are in the
+   * original list, deleting [3:30 - 5:30] will result in two new ranges: [3 - 3:30] and [5:30 - 6].
+   */
+  public void deleteTimeRange(TimeRange timeRangeToDelete) {
+
+    ListIterator<TimeRange> iterator = allTimeRanges.listIterator();
+
+    while (iterator.hasNext()) {
+      TimeRange currentRange = iterator.next();
+      if (currentRange.overlaps(timeRangeToDelete)) {
```

Comment:
It might be simpler to continue in the opposite case, so you don't have to indent the inner block.

Reply:
Done.